File: src/commands/group.ts
/**
 * Project group (folder) commands.
 * 
 * In TickTick's API, "project groups" are what users see as "folders" in the UI.
 * This module provides both `group` and `folder` command aliases.
 */
import { Command } from "commander";
import { getClient } from "../api/client.js";
import {
  printError,
  printSuccess,
  printInfo,
  printJson,
  printKeyValue,
  printGroupsTable,
  printProjectsTableSimple,
} from "../output/index.js";
import { handleError } from "./errors.js";
import { getGlobalOptions } from "../index.js";

/**
 * Build the group/folder command with all subcommands.
 * @param commandName - "group" or "folder"
 */
function buildGroupCommand(commandName: string): Command {
  const isFolder = commandName === "folder";
  const entityName = isFolder ? "folder" : "group";
  const entityNameCap = isFolder ? "Folder" : "Group";
  
  const description = isFolder 
    ? "Manage folders (project groups)" 
    : "Manage project groups (folders)";
  const group = new Command(commandName).description(description);

  // list command
  group
    .command("list")
    .description(`List all ${entityName}s`)
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });
        const groups = await client.getProjectGroups();

        if (options.json) {
          printJson(groups);
        } else if (groups.length === 0) {
          printInfo(`No ${entityName}s found`);
        } else {
          printGroupsTable(groups);
        }
      })
    );

  // add command
  group
    .command("add <name>")
    .description(`Create a new ${entityName}`)
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, name: string, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        const groupData = { name };
        const createdGroup = await client.createProjectGroup(groupData);

        if (options.json) {
          printJson(createdGroup);
        } else {
          printSuccess(`Created ${entityName}: ${createdGroup.name}`);
          printInfo(`ID: ${createdGroup.id}`);
        }
      })
    );

  // edit command
  group
    .command("edit <id>")
    .description(`Edit an existing ${entityName}`)
    .option("-n, --name <name>", "New name")
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, id: string, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Find the group first
        const groups = await client.getProjectGroups();
        const foundGroup = groups.find(
          (g) => g.id === id || g.id?.startsWith(id)
        );

        if (!foundGroup) {
          printError(`${entityNameCap} not found: ${id}`);
          process.exit(1);
        }

        const updateData: Parameters<typeof client.updateProjectGroup>[0] = {
          id: foundGroup.id,
        };

        if (options.name) {
          updateData.name = options.name;
        }

        const updatedGroup = await client.updateProjectGroup(updateData);

        if (options.json) {
          printJson(updatedGroup);
        } else {
          printSuccess(`Updated ${entityName}: ${foundGroup.name}`);
        }
      })
    );

  // delete command
  group
    .command("delete <id>")
    .description(`Delete a ${entityName}`)
    .option("-f, --force", "Skip confirmation")
    .action(
      handleError(async function (this: Command, id: string) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Find the group first
        const groups = await client.getProjectGroups();
        const foundGroup = groups.find(
          (g) => g.id === id || g.id?.startsWith(id)
        );

        if (!foundGroup) {
          printError(`${entityNameCap} not found: ${id}`);
          process.exit(1);
        }

        await client.deleteProjectGroups([foundGroup.id]);
        printSuccess(`Deleted ${entityName}: ${foundGroup.name}`);
      })
    );

  // show command - display folder details and its projects
  group
    .command("show <id>")
    .description(`Show ${entityName} details and its projects`)
    .option("--json", "Output as JSON")
    .action(
      handleError(async function (this: Command, id: string, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });

        // Find the group
        const groups = await client.getProjectGroups();
        const foundGroup = groups.find(
          (g) => g.id === id || g.id?.startsWith(id)
        );

        if (!foundGroup) {
          printError(`Folder not found: ${id}`);
          process.exit(1);
        }

        // Get projects in this folder
        const allProjects = await client.getProjects();
        const folderProjects = allProjects.filter(
          (p) => p.groupId === foundGroup.id
        );

        if (options.json) {
          printJson({
            folder: foundGroup,
            projects: folderProjects,
          });
        } else {
          printKeyValue(
            {
              ID: foundGroup.id ?? "-",
              Name: foundGroup.name ?? "-",
              "Project Count": String(folderProjects.length),
            },
            ["ID", "Name", "Project Count"]
          );

          if (folderProjects.length > 0) {
            console.log();
            printInfo("Projects in this folder:");
            printProjectsTableSimple(folderProjects);
          } else {
            console.log();
            printInfo("No projects in this folder");
          }
        }
      })
    );

  return group;
}

/**
 * Create the "group" command.
 */
export function createGroupCommand(): Command {
  return buildGroupCommand("group");
}

/**
 * Create the "folder" command (alias for group).
 */
export function createFolderCommand(): Command {
  return buildGroupCommand("folder");
}

File: ticktick_cli/cli/group.py
"""Project group commands for TickTick CLI."""

from typing import Annotated

import typer

from ticktick_cli import output
from ticktick_cli.client import AuthenticationError, ClientError, get_client

app = typer.Typer(help="Manage project groups")


def handle_client_error(func):
    """Decorator to handle client errors."""
    from functools import wraps

    @wraps(func)
    def wrapper(*args, **kwargs):
        try:
            return func(*args, **kwargs)
        except AuthenticationError as e:
            output.print_error(str(e))
            raise typer.Exit(1)
        except ClientError as e:
            output.print_error(str(e))
            raise typer.Exit(1)

    return wrapper


@app.command("list")
@handle_client_error
def list_groups(
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """List all project groups."""
    client = get_client()
    groups = client.get_project_groups()

    if json_output:
        output.print_json(groups)
    elif not groups:
        output.print_info("No project groups found")
    else:
        table = output.create_table("ID", "Name")
        for group in groups:
            table.add_row(
                output.truncate_id(group.get("id", "")),
                group.get("name", ""),
            )
        output.print_table(table)


@app.command("add")
@handle_client_error
def add_group(
    name: Annotated[str, typer.Argument(help="Group name")],
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """Create a new project group."""
    client = get_client()
    group = client.create_project_group(name=name)

    if json_output:
        output.print_json(group)
    else:
        output.print_success(f"Created group: {group.get('id', 'unknown')}")


@app.command("edit")
@handle_client_error
def edit_group(
    group_id: Annotated[str, typer.Argument(help="Group ID")],
    name: Annotated[
        str | None,
        typer.Option("--name", "-n", help="New name"),
    ] = None,
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """Edit an existing project group."""
    client = get_client()

    kwargs = {}
    if name:
        kwargs["name"] = name

    if not kwargs:
        output.print_error("No fields to update")
        raise typer.Exit(1)

    group = client.update_project_group(group_id, **kwargs)

    if json_output:
        output.print_json(group)
    else:
        output.print_success(f"Updated group: {group_id}")


@app.command("delete")
@handle_client_error
def delete_group(
    group_ids: Annotated[list[str], typer.Argument(help="Group ID(s) to delete")],
    force: Annotated[
        bool,
        typer.Option("--force", "-f", help="Skip confirmation"),
    ] = False,
) -> None:
    """Delete project group(s)."""
    if not force:
        confirm = typer.confirm(f"Delete {len(group_ids)} group(s)?")
        if not confirm:
            output.print_info("Cancelled")
            return

    client = get_client()
    client.delete_project_groups(group_ids)
    output.print_success(f"Deleted {len(group_ids)} group(s)")
